### PR TITLE
Update kube-openapi

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Updates kube-openapi to latest revision & updates kube-aggregator open-apiv3 code to respond to those changes.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/2898
```

### CODE REVIEW

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/downloader_test.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"k8s.io/kube-openapi/pkg/handler3"
```
Comment (jiahuif) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/downloader_test.go:26]: nit: import grouping

Comment (alexzielenski) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/downloader_test.go:26]: @jiahuif I am not sure what this comment refers to. This ordering is determined by `gofmt`. Do you mean to I should add another newline? or could you show explicitly what you are looking for?

Comment (jiahuif) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/downloader_test.go:26]: Yes, the convention is
import (
standard libraries

third-party libraries

k8s.io dependencies
)

Comment (jiahuif) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/downloader_test.go:26]:
```suggestion

	"k8s.io/kube-openapi/pkg/handler3"
```

**Comment by apelisse** [comment]
Looks good to me, thank you
/approve
/lgtm

**Comment by jiahuif** [comment]
There is some issue with the verify-*.sh scripts. https://github.com/kubernetes/kubernetes/pull/108901

File: hack/update-openapi-spec.sh [hunk 2]
```diff
@@ -107,7 +107,7 @@ mkdir -p "${OPENAPI_ROOT_DIR}/v3"
 # ".well-known__openid-configuration_openapi.json"
 rm -r "${OPENAPI_ROOT_DIR}"/v3/{*,.*} || true
 
-curl -w "\n" -kfsS -H 'Authorization: Bearer dummy_token' "https://${API_HOST}:${API_PORT}/openapi/v3" | jq '.Paths' | jq -r '.[]' | while read -r group; do
+curl -w "\n" -kfsS -H 'Authorization: Bearer dummy_token' "https://${API_HOST}:${API_PORT}/openapi/v3" | jq -r '.Paths | to_entries | .[].key' | while read -r group; do
```
Comment (liggitt) [inline hack/update-openapi-spec.sh:110]: curious why this change was needed

File: vendor/k8s.io/kube-openapi/pkg/handler3/handler.go [hunk 3]
```diff
@@ -49,9 +52,20 @@ const (
 	subTypeJSON     = "json"
 )
 
+// OpenAPIV3Discovery is the format of the Discovery document for OpenAPI V3
+// It maps Discovery paths to their corresponding URLs with a hash parameter included
+type OpenAPIV3Discovery struct {
+	Paths map[string]OpenAPIV3DiscoveryGroupVersion
```
Comment (liggitt) [inline vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:58]: if this is serializing to json is has to define json tags and should follow API conventions of being lower-cased (e.g. `paths` and `url`)

File: vendor/k8s.io/kube-openapi/pkg/handler3/handler.go [hunk 4]
```diff
@@ -49,9 +52,20 @@ const (
...
+}
+
+// OpenAPIV3DiscoveryGroupVersion includes information about a group version and URL
+// for accessing the OpenAPI. The URL includes a hash parameter to support client side caching
+type OpenAPIV3DiscoveryGroupVersion struct {
+	URL string
```
Comment (liggitt) [inline vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:64]: is this a full URL (including hostname/port) or a relative path? if a path, is it a request-relative path or an absolute path?

File: vendor/k8s.io/kube-openapi/pkg/handler3/handler.go [hunk 5]
```diff
@@ -192,7 +228,26 @@ func (o *OpenAPIService) HandleGroupVersion(w http.ResponseWriter, r *http.Reque
...
 				return
 			}
-			w.Header().Set("Etag", etag)
+			// ETag must be enclosed in double quotes: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag
+			w.Header().Set("Etag", strconv.Quote(etag))
+
+			if hash := r.URL.Query().Get("hash"); hash != "" {
+				if hash != etag {
+					u := constructURL(group, etag)
+					http.Redirect(w, r, u, 301)
```
Comment (liggitt) [inline vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:237]: we won't serve a request without a hash, but require a redirect and a second request?

Comment (liggitt) [reply vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:237]: oh, I misread... this is redirecting a request with a stale hash to the current hash

File: hack/update-openapi-spec.sh (hunk 2 above)
Comment (alexzielenski) [reply hack/update-openapi-spec.sh:110]: The format of `/openapi/v3` has changed. The change mirrors that of the removed `gvList` you commented on later in this PR.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/downloader.go [hunk 6]
```diff
@@ -43,11 +44,6 @@ func (s *Downloader) handlerWithUser(handler http.Handler, info user.Info) http.
 	})
 }
 
-// gvList is a struct for the response of the /openapi/v3 endpoint to unmarshal into
-type gvList struct {
-	Paths []string `json:"Paths"`
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/downloader.go:48]: afaik this has always been serialized as `Paths`.

@Jefftree might have more context

File: vendor/k8s.io/kube-openapi/pkg/handler3/handler.go (hunk 3 above)
Comment (alexzielenski) [reply vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:58]: @Jefftree Jordan's comments here in the `vendor` directory I think are more targeted towards your recent PR:

https://github.com/kubernetes/kube-openapi/pull/284

File: vendor/k8s.io/kube-openapi/pkg/handler3/handler.go [hunk 7]
```diff
@@ -207,16 +262,28 @@ func (o *OpenAPIService) RegisterOpenAPIV3VersionedService(servePath string, han
...
+		}
+		return ToV3ProtoBinary(json)
 	})
-
-	o.jsonCache = o.jsonCache.New(func() ([]byte, error) {
-		return specBytes, nil
+	// TODO: This forces a json marshal of corresponding group-versions.
+	// We should look to replace this with a faster hashing mechanism.
+	o.etagCache = o.etagCache.New(func() ([]byte, error) {
+		json, err := o.jsonCache.Get()
```
Comment (liggitt) [inline vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:282]: is it correct to pin the etag to the json serialization (for the same etag to be returned to two different requests with two different body content)

File: vendor/k8s.io/kube-openapi/pkg/handler3/handler.go [hunk 8]
```diff
@@ -96,10 +122,17 @@ func (o *OpenAPIService) getGroupBytes() ([]byte, error) {
 	}
 
 	sort.Strings(keys)
-	group := make(map[string][]string)
-	group["Paths"] = keys
-
-	j, err := json.Marshal(group)
+	discovery := &OpenAPIV3Discovery{Paths: make(map[string]OpenAPIV3DiscoveryGroupVersion)}
+	for gvString, groupVersion := range o.v3Schema {
+		etagBytes, err := groupVersion.etagCache.Get()
```
Comment (liggitt) [inline vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:127]: This is assuming the kube-aggregator method of computing the etag will always match aggregated servers' method of computing etag (because we're embedding etags into url params in the root discovery doc, but then requiring them to match in the actual openapi handler). That seems fragile, and means we can't change how etags are computed in the future.

File: vendor/k8s.io/kube-openapi/pkg/handler3/handler.go [hunk 9]
```diff
@@ -96,10 +122,17 @@ func (o *OpenAPIService) getGroupBytes() ([]byte, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		discovery.Paths[gvString] = OpenAPIV3DiscoveryGroupVersion{
+			URL: constructURL(gvString, string(etagBytes)),
```
[the remark this answers is not in the document]
Comment (apelisse) [reply vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:132]: Yeah, unfortunately it does eliminate some of the benefits from splitting. We're still doing this "lazily" though (so not until someone requests the root document). Jeff has added a comment somewhere to describe that. We're planning on using the swagger->gnostic->pb (the opposite of Alex's PR) to do that much much faster.

File: vendor/k8s.io/kube-openapi/pkg/handler3/handler.go (hunk 8 above)
Comment (Jefftree) [reply vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:127]: For OpenAPI V3, the aggregator will proxy requests to aggregated servers' rather than downloading and recomputing etags/etc.

Comment (liggitt) [reply vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:127]: ah, ok, I missed that

Comment (liggitt) [reply vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:127]: oh, and it will just aggregate their discovery docs, including whatever hashes the backends include?

File: vendor/k8s.io/kube-openapi/pkg/handler3/handler.go (hunk 4 above)
Comment (Jefftree) [reply vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:64]: Any suggestions on naming? I've opened https://github.com/kubernetes/kube-openapi/pull/295

Comment (liggitt) [reply vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:64]: `Path` seems most accurate to me

```
Path string `json:"path"`
```

Comment (liggitt) [reply vendor/k8s.io/kube-openapi/pkg/handler3/handler.go:64]: actually, I guess it includes a query param, so it's not *just* a path :-/

**Comment by alexzielenski** [comment]
pr is updated with @Jefftree's latest changes to kube-openapi

**Comment by liggitt** [comment]
#108906 flake

/retest
/approve

**Comment by Jefftree** [comment]
/retest
Same flake

**Comment by jiahuif** [comment]
/ok-to-test
tide/merge-method-squash

**Comment by jiahuif** [comment]
/test 
pull-kubernetes-e2e-gce-100-performance

**Comment by Jefftree** [comment]
/lgtm
(rebase)
